### PR TITLE
add option to pseudonymizer to use different encryption modes

### PR DESCRIPTION
**Pull request by ekneg54**
This adds an option to the pseudonymizer to choose between "ctr" and "gcm" mode for the aes encryption part of the pseudonymization process.

* use AES in GCM mode to decouple key length of depseudo and analyst keys

TODO:

- [x] add first test
- [x] add implementation for pseudonymization
- [x] add implementation for depseudonymization
- [x] increase documentation
- [x] update changelog

### CODE REVIEW

File: logprep/util/pseudo/encrypter.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    """Hybrid encryption that encrypts a string with AES in GCM Mode and dual PKCS#1"""
+
+    def encrypt(self, input_str: str) -> str:
+        """Encrypt a string using hybrid encryption.
+
+        The input string is encrypted with AES in GCM mode using a random
+        session key. The session key is then encrypted with a depseudo AES key
+        in GCM mode. The depseudo key is then encrypted with the public key
+        of the analyst and the encrypted session key is encrypted with the
+        public key of the depseudonymizer personel. This decouples the analyst
```
Comment (dtrai2) [inline logprep/util/pseudo/encrypter.py:50]:
```suggestion
        public key of the depseudonymizer personnel. This decouples the analyst
```

File: logprep/util/pseudo/encrypter.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        """
+        if not self._pubkey_analyst or not self._pubkey_depseudo:
+            raise ValueError("Cannot encrypt because public keys are not loaded")
+
+        # encrypt input string with AES session key
+        session_key: bytes = get_random_bytes(16)
+        aes_key_input_str = AES.new(session_key, AES.MODE_GCM)
+        input_str_enc: bytes = aes_key_input_str.encrypt(input_str.encode("utf-8"))
+
+        # encrpyt session key with AES depseudo key
```
Comment (dtrai2) [inline logprep/util/pseudo/encrypter.py:71]:
```suggestion
        # encrypt session key with AES depseudo key
```

File: logprep/util/pseudo/encrypter.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+        session_key = get_random_bytes(16)
+
+        cipher_rsa_analyst = PKCS1_OAEP.new(self._pubkey_analyst)
+        enc_session_key = cipher_rsa_analyst.encrypt(session_key)
+
+        cipher_rsa_depseudo = PKCS1_OAEP.new(self._pubkey_depseudo)
+        enc_session_key = cipher_rsa_depseudo.encrypt(enc_session_key)
+
+        cipher_aes = AES.new(session_key, AES.MODE_CTR)  # nosemgrep
```
Comment (dtrai2) [inline logprep/util/pseudo/encrypter.py:122]:
```suggestion
        cipher_aes = AES.new(session_key, AES.MODE_CTR)
```

File: tests/unit/util/pseudo/test_decrpyter.py [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+        assert encrypted_value != "1"
+        # split and decode encrypted value
+        encrypted_value: bytes = base64.b64decode(encrypted_value)
+        session_key_enc_enc = encrypted_value[:128]
+        aes_key_depseudo_nonce = encrypted_value[128:144]
+        depseudo_key_enc = encrypted_value[144:400]
+        aes_key_input_str_nonce = encrypted_value[400:416]
+        input_str_enc = encrypted_value[416:]
+
+        # decrypt encrypted encrypted session key with analyst private key
```
Comment (dtrai2) [inline tests/unit/util/pseudo/test_decrpyter.py:67]:
```suggestion
        # decrypt encrypted session key with analyst private key
```

File: logprep/util/pseudo/decrypter.py [hunk 5]
```diff
@@ -66,53 +130,91 @@ def ciphertext(self) -> bytes:
...
 
         Returns
         -------
-        PKCS1OAEP_Cipher
-            returns a PKCS1OAEP_Cipher representation of the depseudo key
+        bytes
+            All bytes after the first 18 bytes
         """
-        return self._depseudo_key
+        return self.pseudonymized_string[416:]
```
[the remark this answers is not in the document]
Comment (ekneg54) [reply logprep/util/pseudo/decrypter.py:171]: done

**Review by dtrai2 — APPROVED** [review]

File: logprep/util/pseudo/decrypter.py [hunk 6]
```diff
@@ -66,53 +130,91 @@ def ciphertext(self) -> bytes:
...
-    def depseudo_key(self) -> PKCS1OAEP_Cipher:
-        """getter for depseudo_key
+    def ciphertext(self) -> bytes:
+        """the cipher text
 
         Returns
         -------
-        PKCS1OAEP_Cipher
-            returns a PKCS1OAEP_Cipher representation of the depseudo key
+        bytes
```
[the remark this answers is not in the document]
Comment (ekneg54) [reply logprep/util/pseudo/decrypter.py]: done

File: logprep/util/pseudo/encrypter.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+        str
+            String representation of the Base64-encoded concatenation of the
+            double-encrypted session key, the depseudo key AES nonce, the encrypted
+            depseudo key, the AES nonce of the session key and the AES ciphertext
+        """
+        if not self._pubkey_analyst or not self._pubkey_depseudo:
+            raise ValueError("Cannot encrypt because public keys are not loaded")
+
+        # encrypt input string with AES session key
+        session_key: bytes = get_random_bytes(16)
```
[the remark this answers is not in the document]
Comment (ekneg54) [reply logprep/util/pseudo/encrypter.py]: now its in the `logprep.util.defaults` module as constant

File: logprep/util/pseudo/encrypter.py [hunk 8]
```diff
@@ -82,13 +83,13 @@ def encrypt(self, input_str: str) -> str:
...
-            + aes_key_input_str.nonce
-            + input_str_enc
-        ).decode("ascii")
+        return (
+            f"{base64.b64encode(session_key_enc_enc).decode('ascii')}:"
+            f"{base64.b64encode(aes_key_depseudo.nonce).decode('ascii')}:"
+            f"{base64.b64encode(depseudo_key_enc).decode('ascii')}:"
+            f"{base64.b64encode(aes_key_input_str.nonce).decode('ascii')}:"
+            f"{base64.b64encode(input_str_enc).decode('ascii')}"
+        )
```
Comment (dtrai2) [inline logprep/util/pseudo/encrypter.py:92]: how about you implement this as `__str__` or `__repr__` method inside the `GCMPseudonym` dataclass?

File: logprep/util/pseudo/encrypter.py [hunk 9]
```diff
@@ -122,4 +123,8 @@ def encrypt(self, input_str: str) -> str:
         cipher_aes = AES.new(session_key, AES.MODE_CTR)
         ciphertext = cipher_aes.encrypt(input_str.encode("utf-8"))
 
-        return base64.b64encode(enc_session_key + cipher_aes.nonce + ciphertext).decode("ascii")
+        return (
+            f"{base64.b64encode(enc_session_key).decode('ascii')}:"
+            f"{base64.b64encode(cipher_aes.nonce).decode('ascii')}:"
+            f"{base64.b64encode(ciphertext).decode('ascii')}"
+        )
```
Comment (dtrai2) [inline logprep/util/pseudo/encrypter.py:130]: same as above

**Review by dtrai2 — CHANGES_REQUESTED** [review]

**Review by dtrai2 — APPROVED** [review]

**Pull request merged**